### ISSUE TITLE
AMF doesn't check security mode

### ISSUE THREAD

**Issue by mrlnc** [free5gc/free5gc#60]
Hi everybody!

Looks like AMF only checks SecurityHeaderType, but not if the security mode matches the UE context. If a UE already performed the Security Mode procedure, then all incoming NAS messages from that UE must be integrity protected and ciphered.

An incoming NAS message with `SecurityHeaderTypePlainNas` should be discarded. However, AMF seems to accept the message.

### Steps to Reproduce

Apply this patch: https://github.com/mrlnc/free5gc/commit/6fbe8fcb8cf75f5ce907eb8897de296b5ad2a314

The patch will just make the UE component send plain messages instead of properly encrypting and integrity protecting them.

Then run the registration test:
```
$ ./test.sh TestRegistration
```

Despite the AMF configured `NEA2` and `NIA2` in Security Mode Command, it accepted all incoming `PlainNas` messages.

### PCAPs

TestRegistration without modification:
[amf_security_bug_encrypted.pcapng.zip](https://github.com/free5gc/free5gc/files/4718695/amf_security_bug_encrypted.pcapng.zip)

TestRegistration with PLAIN messages (patch applied):
[amf_security_bug_plain.pcapng.zip](https://github.com/free5gc/free5gc/files/4718694/amf_security_bug_plain.pcapng.zip)

* Security Mode Command: `NEA2` and `NIA2` set
* Security Mode Complete incoming from UE: `PLAIN` without integrity or ciphering. This is the first message that should not have been accepted by the network.

### Walkthrough

(just my notes how the incoming NAS message is processed)

1) It reaches `HandleUplinkNasTransport` in `ngap::handler.go`.
2) It's handed to `HandleNAS` in `nas::handler.go`, which will try to decrypt the NAS message:
```golang
msg, err = nas_security.Decode(ue.AmfUe, nas.GetSecurityHeaderType(nasPdu)&0x0f, nasPdu)
```
3) `Decode` in `nas_security::security.go` then actually checks message integrity and encryption -- but only depends on the security header, not the UE security context.
```golang
func Decode(ue *context.AmfUe, securityHeaderType uint8, payload []byte) (msg *nas.Message, err error) {
	integrityProtected := false
	newSecurityContext := false
	ciphering := false
...
switch securityHeaderType {
	case nas.SecurityHeaderTypePlainNas:
	case nas.SecurityHeaderTypeIntegrityProtected:
		integrityProtected = true
	...
}
...
	if ue.CipheringAlg == context.ALG_CIPHERING_128_NEA0 {
		ciphering = false
	}
	if ue.IntegrityAlg == context.ALG_INTEGRITY_128_NIA0 {
		integrityProtected = false
	}
	if ciphering || integrityProtected {
            ...
            /* decryption and integrity check here */
        }
        /* this is always reached */
        err = msg.PlainNasDecode(&payload)
```

The logic should be the other way round: AMF needs to perform security checks by default, and only skip security checks if it was explicitly configured (emergency registration) or not yet enabled (prior security mode procedure).

**Comment by mrlnc** [comment]
I think this should be tagged as bug, not enhancement.

Without enforcing integrity check, anybody can connect to the network without SIM card.

**Comment by free5gc-org** [comment]
That's right.

We're now working on this, it will be fix in our next patch.

Thanks a lot!

**Comment by mrlnc** [comment]
Sure -- just a good opportunity to start working with your code base :)

Some simple change I had in mind:
https://github.com/mrlnc/amf/commit/71407aa15c0a8f814029cfee5b9c0f05f4420584

However I'm not 100% through all your UE context handling. Like, if `ue.SecurityContextAvailable` is reliable to signal the UE security context.

**Comment by Mkishore1** [comment]
Hello all, 
When I am Trying to build network components after "./install_env.sh"  step 

when i run the command 

~/go/src/free5gc$ sudo ./build.sh 

Start build amf…
./build.sh: line 7: go: command not found
Start build ausf…
./build.sh: line 7: go: command not found
Start build nrf…
./build.sh: line 7: go: command not found
Start build nssf…
./build.sh: line 7: go: command not found
Start build pcf…
./build.sh: line 7: go: command not found
Start build smf…
./build.sh: line 7: go: command not found
Start build udm…
./build.sh: line 7: go: command not found
Start build udr…
./build.sh: line 7: go: command not found
Start build n3iwf…
./build.sh: line 7: go: command not found
Start build UPF… CMake Error: The source directory “/home/crrao-aimscs/go/src/free5gc/src/upf” does not appear to contain CMakeLists.txt. Specify --help for usage, or press the help button on the CMake GUI. make: *** No targets specified and no makefile found. Stop.

Thank you!!

**Comment by endrigoshi** [comment]
Did you clone and update all the submodules prior to that?
```bash
cd $GOPATH/src/free5gc
git checkout master
git submodule sync
git submodule update --init --jobs `nproc`
git submodule foreach git checkout master
git submodule foreach git pull --jobs `nproc`
```

**Comment by Mkishore1** [comment]
Thanks,
I followed the above steps and re-run the steps 
I got the following error 
~/go/src/free5gc$ ./build.sh
Start build amf....
WORK=/tmp/go-build634930480
Start build ausf....
WORK=/tmp/go-build972376490
Start build nrf....
WORK=/tmp/go-build106685109
Start build nssf....
WORK=/tmp/go-build028126791
Start build pcf....
WORK=/tmp/go-build212894770
Start build smf....
WORK=/tmp/go-build203129800
Start build udm....
WORK=/tmp/go-build015767116
Start build udr....
WORK=/tmp/go-build160776954
Start build n3iwf....
WORK=/tmp/go-build805855839
Start build UPF....
CMake Error: The source directory "/home/crrao-aimscs/go/src/free5gc/src/upf" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
make: *** No targets specified and no makefile found.  Stop.

it is not able to pick UPF from cmake file 

thank you

**Comment by mrlnc** [comment]
Could you please move your discussion to e.g. the forum or mail?

This GitHub Issue keeps track of the discussion to a very specific issue.

**Comment by Mkishore1** [comment]
Really you people don't give proper responses in the forum!! also it has time limit for each post !!
thank you .

**Comment by free5gc-org** [comment]
Dear all,

We have correct NAS security encode/decode procedure in free5gc v3.0.3. Please check and try it!

Thanks for your report!

**Comment by mrlnc** [comment]
First tests seem very good. The code seems correct now. I added my test here: ac71525def0bb4e4cecc0fde3dd4f2fdb4883bc3 Maybe it's worth to expand this test and include it.

As a side note, I felt like sometimes the testcases fail "too late". So I added some basic checks if the **expected** message was **actually received**. NGAP-level check:
```
assert.True(t, ngapPdu.Present == ngapType.NGAPPDUPresentSuccessfulOutcome && ngapPdu.SuccessfulOutcome.ProcedureCode.Value == ngapType.ProcedureCodeNGSetup, "No NGSetupResponse received.")
```

The same on NAS-layer; check if the expected:
```
assert.True(t, nasPdu.GmmHeader.GetMessageType() == nas.MsgTypeAuthenticationRequest, "No Authentication Request received.")
```

I think UE-side NAS decryption is missing to expand this for other messages :-) I'll see if I can add this myself.
